### PR TITLE
OAK-9535: Support recovery of large branch merge - backport to 1.8

### PR DESCRIPTION
**Pull request by rishabhdaim**
Squashed commits by Stefan Egli and Marcel Reutegger

### CODE REVIEW

**Comment by rishabhdaim** [comment]
@mreutegg @stefan-egli could you please review this PR?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java [hunk 1]
```diff
@@ -130,6 +130,15 @@ public NodeState merge(@NotNull CommitHook hook, @NotNull CommitInfo info)
...
 
+    /**
+     * For test purposes only!
+     * <p>
+     * Forces the branch to persist the changes to the underlying store.
+     */
+    void persist() {
+        branchState.persist();
+    }
+
```
Comment (mreutegg) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java]: Can you please move this method further down and place it between `getMergeLock()` and `merge0()`? This matches the location in trunk and reduces potential conflicts in the future when further changes a ported back.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/TestUtils.java [hunk 2]
```diff
@@ -93,6 +93,14 @@ public static DocumentNodeState asDocumentState(NodeState state){
...
     }
 
+    public static void persistToBranch(NodeBuilder builder) {
+        if (builder instanceof DocumentRootBuilder) {
+            ((DocumentRootBuilder) builder).persist();
+            return;
+        }
+        fail("Not of type DocumentRootBuilder: " + builder.getClass().getName());
+    }
+
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/TestUtils.java]: Similar as above, can you please move this method between `resetRevisionClockToDefault()` and `disposeQuietly()`?

**Review by mreutegg — CHANGES_REQUESTED** [review]

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/RecoveryTest.java [hunk 3]
```diff
@@ -157,6 +169,69 @@ public void recoverOther() throws Exception {
         assertThat(diff.deleted, containsInAnyOrder("/parent/test/c1"));
     }
 
+    @Test
+    @Ignore(value = "Causes OOM on travis and local as well for 1.8 branch")
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/RecoveryTest.java:173]: Why does this work on trunk and 1.22 branch but not on 1.8? Does it get less memory on 1.8 or need more?

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/RecoveryTest.java:173]: We assign all (1.8, 1.22 & trunk) the same memory while running test cases[1]. I even tried updating the `-Xmx` size to `1 GB`, but it didn't work for `1.8` on my local.

We might have some other optimizations/changes that are helping this test pass on `1.22` & `trunk`.

[1] https://github.com/apache/jackrabbit-oak/blob/trunk/oak-parent/pom.xml#L39

**Comment by reschke** [comment]
I belive this PR can be closed; the plan is to EOL 1.8 in May.

**Pull request closed without merging**